### ISSUE TITLE
How to implement a security token that is passed in the URL path (not via querystring)?

### ISSUE THREAD

**Issue by ricosuave73** [tiangolo/fastapi#2797]
I'm looking to implement security with an API key that is passed in a _path variable_.
FastAPI has support for API keys in a _query parameter_, a _header_ or a _cookie_.

I have looked specifically at the query parameter option, but in all cases the API key  needs a name, which a path parameter doesn't have. It seems that query parameter will not work for a variable in the path.

I have also looked at routers and middleware, but I'm unsure if that's the way to go, if I'm not over-complicating things that way.

The bottom line is that for each call I need to look at the URL and check one of the path parameters there to match a valid API key. What would be the best way to solve that in FastAPI?

**Comment by stevenj** [comment]
@ricosuave73 I haven't tested this, but this is what i would do:
Define the path parameter for the api key in the endpoint.
Define the Security Class something like:
```python
class PathAPIKey(SecurityBase):
    type_ = Field(SecuritySchemeType.apiKey, alias="type")
    in_: str = Field(..., alias="in")
    name: str

class PathSecurityBase(SecurityBase):
    def __init__(
        self, auto_error: bool = True
    ):
        self.model = PathAPIKey(in="Path")
        self.auto_error = auto_error

    async def __call__(
        self, request: Request
    ) -> Optional[MyAuthorizationCredentials]:
        apikey = request.path_params['apikey']
        # Validate it here.

        if not_valid:
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_403_FORBIDDEN, detail="Not authenticated"
                )
            else:
                return None
        return MyAuthorizationCredentials(who=what)
```

Make sure you add this as a dependency of the endpoint.  Because you have full access to the underlying request, you have full access to the path and can make any decisions based on that you like.
It is true that OpenAPI only defines apikeys to be query, header or cookie.  But you could set `in` to `Path` and probably the only thing that will mind is an OpenAPI Validator.  I would wager SwaggerUI would just display it.  So in the proposed solution above Imade a version of `class APIKey(SecurityBase)` which defines the normal APIKey security, and just let `in_` take an arbitrary string and not an enum and use that for the model instead of the normal APIKey Model.

**Comment by ricosuave73** [comment]
Thanks, @stevenj ,

I'm kind of confused though:
> Define the path parameter for the api key in the endpoint.

Do you mean in every endpoint? Because that would kind of beat the purpose...
It would be nice if that part of the endpoint is handled somewhere centrally.

**Comment by tiangolo** [comment]
Thanks for the help here @stevenj ! 👏  🙇 

If you solved the original problem, then you can close this issue @ricosuave73  ✔️

Otherwise, please add a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) that I can copy-paste to replicate what you have and to understand your intention.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
